### PR TITLE
Add AdvancedPS

### PR DESCRIPTION
**Pull request by devmotion**
This PR is the Turing counterpart to https://github.com/TuringLang/AdvancedPS.jl/pull/17. It replaces the traces, containers, and logic in Turing.Core with the implementation in the AdvancedPS PR. Additionally, the resampling methods in AdvancedPS are used.

Tests of SMC and PG passed locally.

### CODE REVIEW

**Review by yebai — COMMENTED** [review]
Great work, @devmotion. I carefully reviewed the PR. I didn't notice anything major. See some minor comments below.

File: src/core/container.jl [hunk 1]
```diff
@@ -1,360 +1,49 @@
...
-            # Reset the accumulator of the log probability in the model so that we can
-            # accumulate log probabilities of variables of other samplers until the next
-            # observation.
-            resetlogp!(p.vi)
-        end
-    end
-
-    # Check if all particles are propagated to the final time point.
-    numdone == n && return true
+(f::TracedModel)() = f.model(f.varinfo, f.sampler)
```
[the remark this answers is not in the document]
Comment (devmotion) [reply src/core/container.jl:16]: See my comment in the AdvancedPS PR: https://github.com/TuringLang/AdvancedPS.jl/pull/17#discussion_r535296019

**Comment by yebai** [comment]
I've created a request to register `AdvancedPS`: https://github.com/JuliaRegistries/General/pull/25755

**Review by cpfiffer — APPROVED** [review]
No issues here either. I think we're good to merge this once AdvancedPS 0.1 goes through (https://github.com/JuliaRegistries/General/pull/25755).

**Pull request merged**